x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java

```diff
@@ -83,65 +83,67 @@ public TransportSubmitAsyncSearchAction(
     @Override
     protected void doExecute(Task submitTask, SubmitAsyncSearchRequest request, ActionListener<AsyncSearchResponse> submitListener) {
         final SearchRequest searchRequest = createSearchRequest(request, submitTask, request.getKeepAlive());
-        AsyncSearchTask searchTask = (AsyncSearchTask) taskManager.register("transport", SearchAction.INSTANCE.name(), searchRequest);
-        searchAction.execute(searchTask, searchRequest, searchTask.getSearchProgressActionListener());
-        searchTask.addCompletionListener(new ActionListener<>() {
-            @Override
-            public void onResponse(AsyncSearchResponse searchResponse) {
-                if (searchResponse.isRunning() || request.isKeepOnCompletion()) {
-                    // the task is still running and the user cannot wait more so we create
-                    // a document for further retrieval
-                    try {
-                        final String docId = searchTask.getExecutionId().getDocId();
-                        // creates the fallback response if the node crashes/restarts in the middle of the request
-                        // TODO: store intermediate results ?
-                        AsyncSearchResponse initialResp = searchResponse.clone(searchResponse.getId());
-                        store.createResponse(docId, searchTask.getOriginHeaders(), initialResp, new ActionListener<>() {
-                            @Override
-                            public void onResponse(IndexResponse r) {
-                                if (searchResponse.isRunning()) {
-                                    try {
-                                        // store the final response on completion unless the submit is cancelled
-                                        searchTask.addCompletionListener(
-                                            finalResponse -> onFinalResponse(searchTask, finalResponse, () -> {})
-                                        );
-                                    } finally {
-                                        submitListener.onResponse(searchResponse);
+        try (var ignored = threadContext.newTraceContext()) {
```

Comment:
Is this needed because we are registering a new task in the next line, which will stay alive once the submit action has returned a response? Is there a way to possibly do this automatically when needed? I can see how this is going to be forgotten in other places or possibly there are already places where the same change is needed which we are not aware of yet?

Reply:
Almost all uses of task registration happen via `registerAndExecute`, which handles the thread content stuff automatically. I've checked all the other places which manage tasks without `registerAndExecute` to make sure they're doing the right thing.

I haven't yet come up with an alternative approach unfortunately. Detecting these cases is straightforward at least, because if you don't start a new context, you get an exception!